### PR TITLE
docs(JSDoc): Improve type information

### PR DESCRIPTION
**Pull request by ExE-Boss**
This improves the type information present in JSDoc comments.

---

Pre‑requisite for #461

### CODE REVIEW

**Comment by mde** [comment]
We're already using JSDoc in lots of places, and that includes type information. Is there some reason not to use that?

**Comment by ExE-Boss** [comment]
The current **JSDoc** type information is incorrect in a lot of places.

For instance, `RethrowCallback`’s `lineno` argument is actually a `number`, and it also takes the `EscapeCallback` as the last argument, which wasn’t documented.

Similarly, the `Template` constructor’s `opts` and `resolveInclude`’s `isDir` arguments are optional.

**Comment by mde** [comment]
Could we fix those problems, rather than introducing a new and competing approach? I'm not super tied to JSDoc, but it's pretty prevalent in the codebase, and it does work when it's done correctly.

File: lib/ejs.js [hunk 1]
```diff
@@ -322,20 +323,21 @@ function includeSource(path, options) {
...
  * @static
  */
 
-function rethrow(err, str, flnm, lineno, esc){
+function rethrow(err, str, filename, lineno, esc) {
   var lines = str.split('\n');
   var start = Math.max(lineno - 3, 0);
   var end = Math.min(lines.length, lineno + 3);
-  var filename = esc(flnm); // eslint-disable-line
+  filename = esc(filename);
```
Comment (mde) [inline lib/ejs.js]: Really not keen on making changes to input params inline — is there not a better way to handle this

**Comment by mde** [comment]
Thanks for doing this!

**Pull request merged**